materialize-boilerplate/stream-encode/parquet_schema.go

```diff
@@ -0,0 +1,241 @@
+	LogicalTypeInterval                         // Extends FIXED_LEN_BYTE_ARRAY, with a length of 12 bytes
+	LogicalTypeUnknown                          // Must always be nil
+)
+
+// makeNode translates a ParquetSchemaElement into an actual parquet schema node.
+func makeNode(e ParquetSchemaElement) schema.Node {
+	repetition := parquet.Repetitions.Required
+	if !e.Required {
+		repetition = parquet.Repetitions.Optional
+	}
+
+	switch e.DataType {
+	case PrimitiveTypeInteger:
+		return schema.NewInt64Node(e.Name, repetition, -1)
+	case PrimitiveTypeNumber:
+		return schema.NewFloat64Node(e.Name, repetition, -1)
+	case PrimitiveTypeBoolean:
+		return schema.NewBooleanNode(e.Name, repetition, -1)
+	case PrimitiveTypeBinary:
+		return schema.NewByteArrayNode(e.Name, repetition, -1)
+	case LogicalTypeString:
+		return schema.Must(schema.NewPrimitiveNodeLogical(
+			e.Name,
+			repetition,
+			schema.StringLogicalType{},
+			parquet.Types.ByteArray,
+			-1,
+			-1,
+		))
+	case LogicalTypeUuid:
+		return schema.Must(schema.NewPrimitiveNodeLogical(
+			e.Name,
+			repetition,
+			schema.UUIDLogicalType{},
+			parquet.Types.FixedLenByteArray,
+			16,
+			-1,
+		))
+	case LogicalTypeJson:
+		return schema.Must(schema.NewPrimitiveNodeLogical(
+			e.Name,
+			repetition,
+			schema.JSONLogicalType{},
+			parquet.Types.ByteArray,
+			-1,
+			-1,
+		))
+	case LogicalTypeDate:
+		return schema.Must(schema.NewPrimitiveNodeLogical(
+			e.Name,
+			repetition,
+			schema.DateLogicalType{},
+			parquet.Types.Int32,
+			-1,
+			-1,
+		))
+	case LogicalTypeTime:
+		return schema.Must(schema.NewPrimitiveNodeLogical(
+			e.Name,
+			repetition,
+			schema.NewTimeLogicalType(true, schema.TimeUnitMicros),
+			parquet.Types.Int64,
+			-1,
+			-1,
+		))
+	case LogicalTypeTimestamp:
+		return schema.Must(schema.NewPrimitiveNodeLogical(
+			e.Name,
+			repetition,
+			schema.NewTimestampLogicalType(true, schema.TimeUnitMicros),
+			parquet.Types.Int64,
+			-1,
+			-1,
+		))
+	case LogicalTypeInterval:
+		return schema.Must(schema.NewPrimitiveNodeLogical(
+			e.Name,
+			repetition,
+			schema.IntervalLogicalType{},
+			parquet.Types.FixedLenByteArray,
+			12,
+			-1,
+		))
+	case LogicalTypeUnknown:
+		return schema.Must(schema.NewPrimitiveNodeLogical(
+			e.Name,
+			repetition,
+			schema.UnknownLogicalType{},
+			parquet.Types.Undefined,
+			-1,
+			-1,
+		))
+	default:
+		panic(fmt.Sprintf("makeNode unknown type: %d", e.DataType))
+	}
+}
+
+func FieldsToParquetSchema(fields []string, collection pf.CollectionSpec) ParquetSchema {
+	out := make(ParquetSchema, 0, len(fields))
+
+	for _, f := range fields {
+		p := collection.GetProjection(f)
+		out = append(out, projToSchemaElement(*p))
+	}
+
+	return out
+}
+
+func projToSchemaElement(p pf.Projection) ParquetSchemaElement {
+	out := ParquetSchemaElement{
+		Name:     p.Field,
+		Required: p.Inference.Exists == pf.Inference_MUST,
```

Comment:
Self-review comment: This isn't exactly right; it needs to also consider the possibility for the existence of a `null` value.

Reply:
Nevermind, I actually accounted for this a little lower in the `if t == "null"` check where it sets `Required` to `false`.